mongo/integration/unified/testrunner_operation_test.go

```diff
@@ -0,0 +1,212 @@
+		case "in_progress":
+			expectedState = session.InProgress
+		case "committed":
+			expectedState = session.Committed
+		case "aborted":
+			expectedState = session.Aborted
+		default:
+			return fmt.Errorf("unrecognized session state type %q", stateStr)
+		}
+
+		if actualState := extractClientSession(sess).TransactionState; actualState != expectedState {
+			return fmt.Errorf("expected session state %q does not match actual state %q", expectedState, actualState)
+		}
+		return nil
+	case "assertSessionPinned":
+		return verifySessionPinnedState(ctx, LookupString(args, "session"), true)
+	case "assertSessionUnpinned":
+		return verifySessionPinnedState(ctx, LookupString(args, "session"), false)
+	case "assertSameLsidOnLastTwoCommands":
+		return verifyLastTwoLsidsEqual(ctx, LookupString(args, "client"), true)
+	case "assertDifferentLsidOnLastTwoCommands":
+		return verifyLastTwoLsidsEqual(ctx, LookupString(args, "client"), false)
+	case "assertSessionDirty":
+		return verifySessionDirtyState(ctx, LookupString(args, "session"), true)
+	case "assertSessionNotDirty":
+		return verifySessionDirtyState(ctx, LookupString(args, "session"), false)
+	case "assertCollectionExists":
+		db := LookupString(args, "databaseName")
+		coll := LookupString(args, "collectionName")
+		return verifyCollectionExists(ctx, db, coll, true)
+	case "assertCollectionNotExists":
+		db := LookupString(args, "databaseName")
+		coll := LookupString(args, "collectionName")
+		return verifyCollectionExists(ctx, db, coll, false)
+	case "assertIndexExists":
+		db := LookupString(args, "databaseName")
+		coll := LookupString(args, "collectionName")
+		index := LookupString(args, "indexName")
+		return verifyIndexExists(ctx, db, coll, index, true)
+	case "assertIndexNotExists":
+		db := LookupString(args, "databaseName")
+		coll := LookupString(args, "collectionName")
+		index := LookupString(args, "indexName")
+		return verifyIndexExists(ctx, db, coll, index, false)
+	default:
+		return fmt.Errorf("unrecognized testRunner operation %q", operation.Name)
+	}
+}
+
+func extractClientSession(sess mongo.Session) *session.Client {
+	return sess.(mongo.XSession).ClientSession()
+}
+
+func verifySessionPinnedState(ctx context.Context, sessionID string, expectedPinned bool) error {
+	sess, err := Entities(ctx).Session(sessionID)
+	if err != nil {
+		return err
+	}
+
+	if isPinned := extractClientSession(sess).PinnedServer != nil; expectedPinned != isPinned {
+		return fmt.Errorf("session pinned state mismatch; expected to be pinned: %v, is pinned: %v", expectedPinned, isPinned)
+	}
+	return nil
+}
+
+func verifyLastTwoLsidsEqual(ctx context.Context, clientID string, expectedEqual bool) error {
+	client, err := Entities(ctx).Client(clientID)
+	if err != nil {
+		return err
+	}
+
+	if len(client.started) < 2 {
+		return fmt.Errorf("client has recorded fewer than two command started events")
+	}
+
+	firstID, err := client.started[0].Command.LookupErr("lsid")
```

Comment:
Should this be `client.started[len(client.started) - 2]` instead? And similar for L149.